### PR TITLE
Stored preferences

### PR DESCRIPTION
**Pull request by itsnaidu**
- DashXClient identify method call added
  - after logging-in
  - when authenticated users launch the app freshly
- GraphQL API was implemented for only Fetch and its Model is added
- UI for the settings screen is implemented

### CODE REVIEW

**Comment by itsnaidu** [comment]
https://user-images.githubusercontent.com/47970992/180923301-205b9162-cd3f-4690-971a-f912ca54c26b.mov

**Comment by itsnaidu** [comment]
https://user-images.githubusercontent.com/47970992/181014716-7b9ae4e5-82ce-443f-90b2-278868c65c18.mov

Added progress view and alerts for success and failure - You can observe the Success Alert in the video.

Error Alert
Simulator Screen Shot - iPhone 13 Pro Max - 2022-07-26 at 18 47 51

**Comment by itsnaidu** [comment]
https://user-images.githubusercontent.com/47970992/182373935-1833087d-b610-4df4-87da-db8b0d565754.mov

File: Models.swift [hunk 1]
```diff
@@ -177,4 +177,27 @@ struct User: Codable {
...
     }
+    
+    var idString: String? {
+        id == nil ? nil : String(id!)
+    }
+}
+
+// MARK: - PreferenceDataResponse
+struct PreferenceDataResponse: Codable {
+    let newBookMark: [String : Bool]
```
Comment (RabidFire) [inline Models.swift]: bookmark is one word, so this should be `newBookmark` - not `newBookMark`

File: Models.swift [hunk 2]
```diff
@@ -177,4 +177,27 @@ struct User: Codable {
...
+    let newPost: [String: Bool]
+    
+    var newBookMarkNotificationEnabled: Bool {
+        newBookMark["enabled"] ?? false
+    }
+    
+    var newPostNotificationEnabled: Bool {
+        newPost["enabled"] ?? false
+    }
```
Comment (RabidFire) [inline Models.swift]: Similar to Android, shouldn't we add a `Preference` class that is exposed by SDK which contains the keys `enabled`, `email`, `sms`, etc?

File: Utils/DashXUtils.swift [hunk 3]
```diff
@@ -10,19 +10,32 @@ import DashX
...
-    
-    static func getClient2() -> DashXClient {
-//        DashX.setPublicKey(to: "R7De1FaPwj6JkG1aDICbO82V")
-//        DashX.setBaseURI(to: "https://api.dashx-staging.com/graphql")
-        
+    static func getClient() -> DashXClient {
+        DashX.setPublicKey(to: "Obmgjd5Ail9L1UBzI7RpPz2Y")
+        DashX.setBaseURI(to: "https://api.dashx-staging.com/graphql")
+        DashX.setTargetEnvironment(to: "staging")
         return DashX
```
Comment (RabidFire) [inline Utils/DashXUtils.swift]: Can we get rid of this DashXUtils class? It feels like this is not showing a good way to integrate with DashX.

File: ViewControllers/MainTabBarController.swift [hunk 4]
```diff
@@ -14,6 +14,8 @@ class MainTabBarController: UITabBarController {
     // MARK: ViewDidLoad
     override func viewDidLoad() {
         super.viewDidLoad()
+        DashXUtils.trackEventClient()
+        DashXUtils.client1.track("TestEventFromiOSApp")
```
Comment (RabidFire) [inline ViewControllers/MainTabBarController.swift]: Let's not have events in the code which aren't also present in other demo apps...

File: Utils/DashXUtils.swift [hunk 5]
```diff
@@ -10,19 +10,32 @@ import DashX
...
-    // Client using explicit instance
-    static let client1 = DashXClient(withPublicKey: "R7De1FaPwj6JkG1aDICbO82V",
-                                    baseURI: "https://api.dashx-staging.com/graphql")
-    
-    static func getClient2() -> DashXClient {
-//        DashX.setPublicKey(to: "R7De1FaPwj6JkG1aDICbO82V")
-//        DashX.setBaseURI(to: "https://api.dashx-staging.com/graphql")
-        
+    static func getClient() -> DashXClient {
+        DashX.setPublicKey(to: "Obmgjd5Ail9L1UBzI7RpPz2Y")
```
Comment (RabidFire) [inline Utils/DashXUtils.swift]: We shouldn't have any keys hard-coded like this.

File: Models.swift (hunk 2 above)
Comment (itsnaidu) [reply Models.swift]: Yeah, we can do that. Will update

File: ViewControllers/MainTabBarController.swift (hunk 4 above)
Comment (itsnaidu) [reply ViewControllers/MainTabBarController.swift]: Sure, will remove these track events

File: Utils/DashXUtils.swift (hunk 5 above)
Comment (itsnaidu) [reply Utils/DashXUtils.swift]: Sure, will move it to constants

File: ViewControllers/LoginViewController.swift [hunk 6]
```diff
@@ -111,6 +112,21 @@ class LoginViewController: UIViewController {
...
+                let userDetails: NSDictionary = [
+                    UserAttributes.UID: user.idString as Any,
+                    UserAttributes.EMAIL: user.email as Any,
+                    UserAttributes.NAME: user.name,
+                    UserAttributes.FIRST_NAME: user.firstName as Any,
+                    UserAttributes.LAST_NAME: user.lastName as Any
+                ]
+                DashX.setIdentity(uid: user.idString)
+                do {
+                    try DashX.identify(withOptions: userDetails)
```
Comment (RabidFire) [inline ViewControllers/LoginViewController.swift]: Just keep setIdentity. No need to save registered user details via SDK as the backend API is already doing that

File: ViewControllers/MoreItemsViewController.swift [hunk 7]
```diff
@@ -45,6 +45,24 @@ class MoreItemsViewController: UIViewController {
         self.navigationController?.pushViewController(updateProfileVC, animated: true)
     }
     
+    func goToSettingsScreen() {
+        let settingsVC = UIViewController.instance(of: SettingsViewController.identifier)
+        self.navigationController?.pushViewController(settingsVC, animated: true)
+    }
+    
+    func showAlertToLogoutConfirmation() {
+        let logoutConfirmationAlert = UIAlertController(title: "Confirm Logout", message: "Please confirm you are logging out", preferredStyle: .alert)
```
Comment (RabidFire) [inline ViewControllers/MoreItemsViewController.swift]: Change message: Are you sure you wish to logout?

File: ViewControllers/SettingsViewController.swift [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+    // MARK: - Outlets
+    @IBOutlet weak var someOneCreatesAPostSwitch: UISwitch!
+    @IBOutlet weak var someOneBookmarksYourPostSwitch: UISwitch!
+    var rightBarButton: UIBarButtonItem!
+    var preferenceData: PreferenceDataResponse?
+    
+    private var newBookmarkNotificationEnabled: Bool = false {
+        didSet {
+            if someOneBookmarksYourPostSwitch != nil {
+                someOneBookmarksYourPostSwitch.isOn = newBookmarkNotificationEnabled
```
Comment (RabidFire) [inline ViewControllers/SettingsViewController.swift]: Someone is one word

File: ViewControllers/LoginViewController.swift [hunk 9]
```diff
@@ -111,6 +112,21 @@ class LoginViewController: UIViewController {
...
             LocalStorage.instance.setToken(response.token)
+            if let user = user {
+                let userDetails: NSDictionary = [
+                    UserAttributes.UID: user.idString as Any,
+                    UserAttributes.EMAIL: user.email as Any,
+                    UserAttributes.NAME: user.name,
+                    UserAttributes.FIRST_NAME: user.firstName as Any,
+                    UserAttributes.LAST_NAME: user.lastName as Any
+                ]
+                DashX.setIdentity(uid: user.idString)
```
Comment (RabidFire) [inline ViewControllers/LoginViewController.swift]: Follow same thing as Android

```
    fun setIdentity(uid: String, token: String) {
        this.accountUid = uid
        this.identityToken = token
        saveToStorage()

        createApolloClient()
    }
```

**Comment by itsnaidu** [comment]
@RabidFire  AppDelegate is still the initial place to configure SDKs since we don't have dependencies on the App `window`.

**Pull request merged**